### PR TITLE
Add data warehouse context provider for inference

### PR DESCRIPTION
**Pull request by serramatutu**
# What does this PR do?
This PR adds an inference context provider that fetches min, max, unique and null count statistics from a given set a tables.

# Why does it do it?
This functionality will be useful for implementing inference rules about that set of tables, e.g, columns ending with "_id" are probably primary or foreign keys.

It is the first step towards data source configuration inference.

# What comes next?
For this to be in a minimally usable state, we still need to do the following (which will come on subsequent PRs):
- Implement basic inference rules such as "`_id` columns are probably IDs". Those rules will produce "signals", contextual pieces of evidence about a table or a column.
- Implement an inference policy that coerces signals about a column into some useful piece of information, such as "this column is an ID"
- Implement an `InferenceRunner` class for tying it all together in a sequence of steps (fetching context, processing rules, coercing it into inferred facts, getting user approval, writing config files)
- Surface it to end users via the CLI.

### CODE REVIEW

File: metricflow/inference/context/data_warehouse.py
Comment (tlento) [inline metricflow/inference/context/data_warehouse.py:102]: If you follow the suggestion above this becomes something like:
```suggestion
                column=SqlColumn(table=table, name=col_name, sql_type=SqlColumnType.from_pandas_dtype(str(series.dtype)),
```

With a factory method it'd be more like:
`SqlColumn.from_pandas_dtype(table=table, name=col_name, pandas_dtype=series.dtype)`

File: metricflow/inference/context/data_warehouse.py
Comment (tlento) [inline metricflow/inference/context/data_warehouse.py:104]: This is going to produce the wrong value. You can't do a distinct on a non-hash sample (and *especially* not on a limit sample, since those rows might be ordered on insert) and make any claims at all. Imagine a table that has 10,000 distinct values in this column 10 of which represent 99.99% of rows, or one with 10,000 evenly distributed values that was insert ordered on that column, or one with a time dependency on a much-more-common time series insertion order. 

In the more common cases, randomly distributed moderate cardinality columns are going to produce very high variance (i.e., completely unreliable) output with this strategy, even when normalized against input row count.

I know this is for a by-design-constraint imperfect inference step, but this seems like it'll produce a lot more noise than we want. I recommend leaving this out altogether for the time being, or else doing it via a separate column sampling pass only for cases where it seems warranted (or where the user requests us to do the relevant set of inferences, or something).

File: metricflow/inference/context/data_warehouse.py
Comment (tlento) [reply metricflow/inference/context/data_warehouse.py:39]: Actually, looking further along, I think we shouldn't have a `cardinality` property at all, unless we can populate it with a reasonable true value.

Another option is to have `cardinality` be an enum of literals like LOW, MEDIUM, and HIGH where we set some sample-driven thresholds

**Review by tlento — COMMENTED** [review]
Neat! Three things:

1. Very high level, I'm not sure if mediating all of our measures through PANDAS is the approach we want. It's hard to do engine-specific rendering for these stats queries today but that's a thing we might want to make easier anyway and I don't think it's very difficult to do. If we were to do that, would you prefer an approach where this was all managed via native SQL to doing an in-memory dump to PANDAS?
2. The sampling strategy here is highly suspect, particularly for the distinct count set, and to a lesser extent for NULL detection. We're going to get a lot of bad values for the former and a lot of false negatives for the latter unless we can routinely load the full table - or at least a substantial fraction of it - into memory.
3. I left a number of code structure comments inline for your consideration. You definitely don't want to take dependencies on ParseableField, it's not useful here.

Let me know if you have any questions!

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py:21]: Yeah, I also thought about that. However, I intended `SqlColumn` to be just a _reference_ to a column. We have no way of getting a column's type without reaching out to the warehouse directly to ask for it (or assume the client knows it).

I'd like to keep `SqlColumn` as just a reference class that doesn't require any requests to be populated.

I do understand that `type` is not exactly a statistic either. What do you think if renaming `ColumnStatistics` to something like `ColumnAttributes` or something?

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py:39]: 👍

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py:104]: **About PANDAS usage:**
I was actually going for the native SQL approach for a couple of reasons:
1. We're not constrained by the client's memory and could theoretically submit all the queries at once to be run natively at the warehouse. I think this could be a neat performance improvement for the future.
2. The data warehouse is built exactly for querying, and is probably much more reliable and fast than we can probably ever get with a naive client-side approach.

I was initially planning on running something like this:
```sql
with source as (
    select * from DEMO_DB.CUSTOMERS limit 10000
    -- possibly some other fancy sampling strategy here
)

select 
    count(*) as count, 
    
    count(distinct customer_id) as customer_id_count,
    min(customer_id) as customer_id_min,
    max(customer_id) as customer_id_max,
    sum(case when customer_id is null then 1 else 0 end) as customer_id_countnull,
    
    count(distinct ds) as ds_count,
    min(ds) as ds_min,
    max(ds) as ds_max,
    sum(case when ds is null then 1 else 0 end) as ds_countnull
from
    source
```

However, I thought warehouses could get kinda finicky with ANSI SQL and for querying stuff like `MIN` for a nested array field, JSON or other weird field. I decided for running it all client-side just to get a thing going for V0. Do you think that's a reasonable decision or should we just switch to native SQL from the start?

**About sampling**
I was hoping with a sufficiently large N, the sample distribution of nulls, different values etc would mimic the real distribution in the data. I guess this is not the case for time series data that's inserted in order. I'm also aware this could be very problematic for MIN/MAX values, since random samples are very likely to get those values right.

I do think we may still want to employ some sampling strategy, though. My main concern with this is running something like `COUNT(DISTINCT user_id)` over a huge table (terabytes or petabytes) will for sure get _very_ expensive.

Honestly, I don't know what is a good solution to this for now.

File: metricflow/dataflow/sql_column.py
Comment (serramatutu) [reply metricflow/dataflow/sql_column.py:43]: K. Didn't know you were removing it :P

File: metricflow/inference/context/data_warehouse.py
Comment (tlento) [inline metricflow/inference/context/data_warehouse.py:97]: I just went through your examples more carefully, and I think it's worth noting:

This sampling approach is not good for anything. You can't even get a viable row count out of it. So I'd definitely get rid of that before doing any real world testing.

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py]: Fixed. Thanks for this!

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py]: I ended up making it clearer in each class' docs what they're for. I also renamed `ColumnStatistics` to `ColumnProperties`, and `TableStatistics` to `TableProperties`.

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py]: Fixed.

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py]: Done.

**Comment by serramatutu** [comment]
You might wonder why I didn't incorporate all the Snowflake-specific SQL and whatnot into the `SqlClient` class. There are mainly 2 reasons for that:
1. This is an experimental feature, and I didn't want to mess with the core of the framework and add room for a thing we don't even know we'll need.
2. For v0 inference is going to be specific to Snowflake. We might want to add a better generalization for this (and move some of the code around) once (and if) we decide to make this a feature for all supported warehouses.

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py]: I'm resolving this since we've moved to a warehouse-native approach, which gets rid of client-side Pandas and uses Snowflake's native `SAMPLE` capabilities.

I'll definitely watch out for noise in the data once we're implementing all the inference rules and policies. I guess the only way to estimate a tolerable amount of noise is to actually run inference over tables and see how this performs.

File: metricflow/test/inference/context/test_snowflake.py
Comment (tlento) [inline metricflow/test/inference/context/test_snowflake.py:92]: This is not a useful assertion. You're checking that a very specific call is being made, but this means we need to modify this whenever we change anything. Like if I decide to change "SUM(CASE...)" to "SUM(IF(....))" because that turns out to be faster or something, now I need to update this test. If I switch the formatting to use a jinja template or something with nicer line breaks, now I need to update this test.
```suggestion
```

File: metricflow/test/inference/context/test_snowflake.py
Comment (tlento) [inline metricflow/test/inference/context/test_snowflake.py:62]: So I'm not a huge fan of mocks in this case. I'd prefer a proper fake, where there was like a local isolated database we could query, but I don't think Python provides that, so the mocking you're doing here might be as close as it gets.

At any rate, this test really only checks the mechanics against the boundary of the Snowflake query API itself, which can be fine, but here I don't think it gets us very much. There isn't very much data manipulation on the results to convert them into the TableProperties and ColumnProperties constructs, and the internal logic of those classes is already tested elsewhere.

It's fine to leave this here, but I think you'll want an integration test at some point. However, that should probably be in the base warehouse class, and it should essentially be using the correct context provider depending on which engine is under test.

**Review by tlento — COMMENTED** [review]
Ok, this makes sense to me, thanks for the updates!

I left a bunch of cleanup notes inline. They largely fall into two categories:

1. API/signature cleanup and refinement (initializer, ABC designation, etc.)
2. Test cleanup (removing unneeded assertions, docstrings, etc.)

Shouldn't be much more to do there. In the long run you'll want an integration test but I think holding off on that during the prototype phase makes sense since you'll need to set it up to skip whenever a not-snowflake engine is invoked (which is most of the time in the rest of the Metricflow dev world).

File: metricflow/inference/context/base.py
Comment (serramatutu) [reply metricflow/inference/context/base.py:14]: Hmmmm. Yeah, I just came up with some names and rolled with it, but happy to change it if there's anything that makes it clearer.

Just for context, in the future I'm planning to add other types of "contexts", such as a `DbtModelContext` that could use data from your DBT models to infer datasource configurations. 

So a "context" is really just a piece of information that comes from somewhere and is useful for inference. The "provider" is just a thing that actually gets that information by making API calls, database calls etc.

I'm trying to think of a better nomenclature but that's the best I could come up with :p

File: metricflow/inference/context/snowflake.py
Comment (serramatutu) [reply metricflow/inference/context/snowflake.py:17]: Yeah that's true. For now I'll leave it like this. If we generalize this we should somehow keep this in the base implementation class.

File: metricflow/dataflow/sql_column.py
Comment (serramatutu) [reply metricflow/dataflow/sql_column.py:15]: Yep. At least for inference they're kind of the same thing: a real number. The precision doesn't really matter I think.

File: metricflow/test/inference/context/test_data_warehouse.py
Comment (serramatutu) [reply metricflow/test/inference/context/test_data_warehouse.py:42]: I guess I overlooked some of the assertions here: I removed `dtype` but forgot to remove the test for it. The only real thing worth testing for `ColumnProperties` is `is_empty`.

For `TableProperties`, I'm testing the list-to-dict that happens in `__post_init__`. I will add this to a docstring.

File: metricflow/test/inference/context/test_data_warehouse.py
Comment (serramatutu) [reply metricflow/test/inference/context/test_data_warehouse.py:154]: 👍

File: metricflow/test/inference/context/test_snowflake.py
Comment (serramatutu) [reply metricflow/test/inference/context/test_snowflake.py:92]: 👍

File: metricflow/inference/context/data_warehouse.py
Comment (serramatutu) [reply metricflow/inference/context/data_warehouse.py:78]: Genuine question: is using `@dataclass` for auto-implementing magic methods a bad practice? From [python docs]()

> This module provides a decorator and functions for automatically adding generated [special method](https://docs.python.org/3/glossary.html#term-special-method)s such as __init__() and __repr__() to user-defined classes.

I was basically using this for adding an `__init__` method to it without the boring work of assign properties to `self`.

**Review by tlento — APPROVED** [review]
Looking good, thanks for the updates!

Please do update the DataWarehouseInferenceContextProvider before final cleanup and merge - we want that abstract method labeled as such, and that means we're better off not using dataclass there. I have a not-very-important rename suggestion inline as well, up to you what to do with that.

No need to ask for additional review unless you want eyes on further updates prior to merge.

File: metricflow/inference/context/base.py
Comment (tlento) [reply metricflow/inference/context/base.py:14]: Naming is so hard. :)

I think adding short form working examples (one with dbt models and one with warehouse metadata or column statistics) to the documentation will help a lot here.

**Pull request merged**
